### PR TITLE
Added a simple request validator

### PR DESCRIPTION
**Pull request by Pchelolo**
A simple request validator based on json-schema validator. Currently, we only validate that all required parameters are present, but the validator is very extensible. Also fixed some tests a bit.

### CODE REVIEW

**Comment by eevans** [comment]
Nice!  Do we know what impact this has on performance?

**Comment by Pchelolo** [comment]
@eevans this code can make ~20 million validations per second on my machine, while I have approximately 1000 req/s, so, simple calculations shows, that the impact is about 0.005 :)

File: lib/validator.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        properties: {}
+    };
+
+    parameters.forEach(function(param) {
+        if (!schema.properties[inMapping[param.in]]) {
+            schema.properties[inMapping[param.in]] = {
+                type: 'object',
+                properties: {}
+            };
+            schema.required = schema.required || [];
```
Comment (gwicke) [inline lib/validator.js]: Would it be okay to initialize required to the empty array right away (around line 24)?

Comment (Pchelolo) [inline lib/validator.js]: @gwicke no, according to the son-schema spec, required array must have at least on element. If we preinitialise it, it wouldn't compile if there's no required fields.

Comment (gwicke) [inline lib/validator.js]: Actually, might be worth mentioning that in a comment, so that we remember later.

File: lib/validator.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            status: 400,
+            body: {
+                type: 'bad_request',
+                title: ajv.errorsText(this._validatorFunc.errors),
+                req: req
+            }
+        });
+    }
+};
+
```
Comment (gwicke) [inline lib/validator.js]: I love how clean this is! Great work!

**Comment by eevans** [comment]
>> @eevans:
>> Do we know what impact this has on performance?
>@Pchelolo:
> this code can make ~20 million validations per second on my machine, while I have approximately 1000 req/s, so, simple calculations shows, that the impact is about 0.005% :)

So, for all intents and purposes... none.  :)  Awesome.

File: lib/restbase.js [hunk 3]
```diff
@@ -283,8 +282,18 @@ RESTBase.prototype._request = function(req) {
...
         // Call the handler with P.try to catch synchronous exceptions.
-        var reqHandlerPromise = P.try(handler, [childRESTBase, childReq])
-        .then(function(res) {
+        var reqHandlerPromise;
+        if (handler.validator) {
+            reqHandlerPromise = P.try(function() {
+                return handler.validator.validate(childReq);
+            })
+            .then(function() {
+                return handler(childRESTBase, childReq);
```
Comment (d00rman) [inline lib/restbase.js]: Shouldn't this be `return P.try(function() {handler(childRESTBase, chlidReq) })` to keep the same functionality as before (catching sync exceptions) ?

Comment (Pchelolo) [inline lib/restbase.js]: @d00rman it's already in the promise, so it's equal to the previous functionality. The `P.try` was needed before to 'initialise' a promise chain.

Comment (d00rman) [inline lib/restbase.js]: Ah right. From the [BB docs](https://github.com/petkaantonov/bluebird/blob/master/API.md#promisetryfunction-fn--arraydynamicdynamic-arguments--dynamic-ctx----promise):

> Start the chain of promises with `Promise.try`. Any synchronous exceptions will be turned into rejections on the returned promise.

**Pull request merged**